File: flutter_readium/ios/flutter_readium/Sources/flutter_readium/model/FlutterMediaOverlay.swift
import Foundation
import ReadiumShared

struct FlutterMediaOverlay {
  let items: [FlutterMediaOverlayItem]
  
  let readingOrderDuration: TimeInterval?
  
  var audioFile: String? {
    items.first?.audioFile
  }
  
  var textFile: String? {
    items.first?.textFile
  }
  
  var totalDuration: TimeInterval? {
    items.reduce(0) { $0 + ($1.audioDuration ?? 0.0) }
  }

  func itemInRangeOfTime(_ time: Double, inHref href: String) -> FlutterMediaOverlayItem? {
    if (href != audioFile && href != textFile) {
      return nil
    }

    return items.first(where: { $0.isAudioInRangeOfTime(time, inHref: href) })
  }
  
  func itemFromTextId(_ textId: String, inHref href: String) -> FlutterMediaOverlayItem? {
    if (textFile != href && audioFile != href) {
      return nil
    }
    
    return items.first(where: { $0.textId == textId })
  }
  
  /// `allowResourceFallback` gates the *imprecise* resource-first fallbacks (an
  /// unmatched text id, or no id at all on an HTML resource): when true, an uncued
  /// anchor maps to the first cue of the resource; when false, returns nil so the
  /// caller can leave playback untouched. Pass false when audio is already playing
  /// in this same resource (cross-resource check is done at the navigator level).
  /// See issue #139.
  func itemFromLocator(_ locator: Locator, allowResourceFallback: Bool = true) -> FlutterMediaOverlayItem? {
    let href = locator.href.string
    if (textFile != href && audioFile != href) {
      return nil
    }

    // Audio time fragment → exact item by time range.
    let timeOffset = locator.timeOffset
    if (timeOffset != nil) {
      return itemInRangeOfTime(timeOffset!, inHref: href)
    }

    // DiViNa comic page: the page locator's only anchor is a generic "img" css selector
    // (which `textId` would otherwise pick up), not a media-overlay text id. Match the
    // page's first narrated item by href — the textref is the image itself.
    if (locator.mediaType.isBitmap) {
      return items.first(where: { $0.textFile == href })
    }

    // Reflowable text: try exact DOM element id match first.
    if let textId = locator.textId, let item = itemFromTextId(textId, inHref: href) {
      return item
    }

    // No id, or id matched no cue → resource-first fallback, gated by policy.
    // Covers: no fragment on an HTML resource, AND an id that has no narration cue
    // (e.g. a ToC entry pointing at a heading that lacks its own sync data).
    if allowResourceFallback && [MediaType.html, MediaType.xhtml].contains(locator.mediaType) {
      if locator.textId != nil {
        Log.navigator.warn("itemFromLocator: textId matched no cue in \(href); falling back to first item in resource")
      }
      return items.first(where: { $0.textFile == href })
    }

    return nil
  }
  
  static func fromJson(_ json: [String: Any], atPosition position: Int, atTocHref: String? = nil, readingOrderDuration: TimeInterval? = nil) -> FlutterMediaOverlay? {
    guard let topNarration = json["narration"] as? [[String: Any]] else { return nil }
    var acc: [FlutterMediaOverlayItem] = []
    
    for obj in topNarration {
      if let item = FlutterMediaOverlayItem.fromJson(obj, atPosition: position, atTocHref: atTocHref, readingOrderDuration: readingOrderDuration) {
        acc.append(item)
      }
      // recurse if nested containers also have "narration"
      if let nested = FlutterMediaOverlay.fromJson(obj, atPosition: position, atTocHref: atTocHref, readingOrderDuration: readingOrderDuration) {
        acc.append(contentsOf: nested.items)
      }
    }
    return FlutterMediaOverlay(items: acc, readingOrderDuration: readingOrderDuration)
  }
}

struct FlutterMediaOverlayItem {
  let audio: String
  let text: String
  let position: Int
  let readingOrderDuration: TimeInterval?
  
  let audioFile: String
  let audioMediaType: MediaType
  private let audioFragment: String

  let audioStart: Double?
  let audioEnd: Double?
  
  var audioDuration: Double? {
    guard let audioStart, let audioEnd else { return nil }
    return max(0, audioEnd - audioStart)
  }
  
  let textFile: String
  let textId: String
  
  let tocTitle: String?
  let tocHref: String?
  
  init(audio: String, text: String, position: Int, tocTitle: String? = nil, tocHref: String? = nil, readingOrderDuration: TimeInterval? = nil) {
    self.audio = audio
    self.text = text
    self.position = position
    self.tocTitle = tocTitle
    self.tocHref = tocHref
    self.readingOrderDuration = readingOrderDuration
    self.audioFile = audio.split(separator: "#", maxSplits: 1).first.map(String.init) ?? audio
    self.audioFragment = audio.split(separator: "#", maxSplits: 1).getOrNil(1).map(String.init) ?? ""
    self.textFile = text.split(separator: "#", maxSplits: 1).first.map(String.init) ?? ""
    self.textId = text.split(separator: "#", maxSplits: 1).getOrNil(1).map(String.init) ?? ""
    self.audioMediaType = switch (audioFile.split(separator: ".").last) {
      case "opus" :
        MediaType.opus
      default:
        MediaType.mpegAudio
    }
    
    if let range = MediaTimeFragment.range(from: audioFragment) {
      self.audioStart = range.start
      self.audioEnd = range.end
    } else {
      self.audioStart = nil
      self.audioEnd = nil
    }
  }
  
  func copyWith(tocTitle: String?, tocHref: String?) -> FlutterMediaOverlayItem {
    return FlutterMediaOverlayItem(audio: audio, text: text, position: position, tocTitle: tocTitle, tocHref: tocHref, readingOrderDuration: readingOrderDuration)
  }
  
  static func == (lhs: FlutterMediaOverlayItem, rhs: FlutterMediaOverlayItem) -> Bool {
    return lhs.audio == rhs.audio && lhs.text == rhs.text && lhs.position == rhs.position
  }
  
  /// Check if this MediaOverlayItem matched href and has time-fragment range matching a given time.
  func isAudioInRangeOfTime(_ time: Double, inHref href: String) -> Bool {
    if (textFile != href && audioFile != href) {
      return false
    }
    guard let start = audioStart else { return false }
    // A reversed or non-finite end (malformed `t=start,end` fragment) would trap
    // `start...end`; treat it as open-ended from `start`, mirroring the no-end case.
    guard let end = audioEnd, end >= start else { return time >= start }
    return (start...end).contains(time)
  }
  
  // MARK: Locators
  
  /// Create a Text-based Locator representing this MediaOverlayItem
  var asTextLocator: Locator? {
    guard
      let href = URL(string: text.split(separator: "#", maxSplits: 1).first.map(String.init) ?? "")
    else { return nil }
    
    let frag = text.split(separator: "#", maxSplits: 1).dropFirst().first.map(String.init)
    var locator = Locator(
      href: href,
      mediaType: MediaType.xhtml,
      title: tocTitle,
      locations: .init(
        fragments: frag.map { [$0] } ?? [],
      )
    )
    if (frag != nil) {
      locator.locations.otherLocations["cssSelector"] = .string("#\(frag!)")
    }
    if (tocHref != nil) {
      locator.locations.otherLocations["tocHref"] = .string(tocHref!)
    }
    return locator
  }
  
  /// Create an Audio-based Locator representing this MediaOverlayItem
  var asAudioLocator: Locator? {
    guard let href = URL(string: audioFile) else { return nil }
    let start = audioStart ?? 0.0
    return Locator(
      href: href,
      mediaType: audioMediaType,
      locations: .init(fragments: [MediaTimeFragment.string(start)])
    )
  }
  
  /// Combine this MediaOverlayItem as a Text-based Locator, with an Audio-based Locator.
  /// This is generally used to report back a synchronizable Locator to Flutter client and backends.
  func toCombinedLocator(fromAudioLocator audioLocator: Locator) -> Locator? {
    guard var textLocator = self.asTextLocator else { return nil }
    // Combine the text-locator with given audio-locator's locations.
    // We keep the otherLocations("cssSelector") from text-locator.
    // We get the position from they MediaOverlay.position
    //
    // Keep the text DOM-id fragment ahead of the audio `t=…` fragment so the combined
    // locator stays self-sufficient for swift-toolkit's reflowable navigator, which
    // positions via `fragments.first` and ignores `cssSelector` (unlike kotlin/ts).
    // `Locator.timeOffset` still finds the `t=` fragment by prefix regardless of order, so
    // audio mapping is unaffected. See docs/parity/locator-field-priority.md.
    let textFragments = textLocator.locations.fragments
    textLocator.locations = Locator.Locations(
      fragments: textFragments + audioLocator.locations.fragments,
      progression: audioLocator.locations.progression,
      totalProgression: audioLocator.locations.totalProgression,
      position: self.position + 1,
      otherLocations: textLocator.locations.otherLocations,
    )
    return textLocator
  }
  
  // MARK: JSON
  static func fromJson(_ json: [String: Any], atPosition position: Int, atTocHref: String?, readingOrderDuration: TimeInterval?) -> FlutterMediaOverlayItem? {
    guard
      let audio = json["audio"] as? String, !audio.isEmpty,
      let text  = json["text"]  as? String, !text.isEmpty
    else { return nil }
    return FlutterMediaOverlayItem(audio: audio, text: text, position: position, tocHref: atTocHref, readingOrderDuration: readingOrderDuration)
  }
}

File: flutter_readium/android/src/main/kotlin/dk/nota/flutterreadium/models/FlutterMediaOverlay.kt
package dk.nota.flutterreadium.models

import android.os.Parcelable
import dk.nota.flutterreadium.PluginLog
import dk.nota.flutterreadium.getTextId
import dk.nota.flutterreadium.progression
import kotlinx.parcelize.IgnoredOnParcel
import kotlinx.parcelize.Parcelize
import org.json.JSONArray
import org.json.JSONObject
import org.readium.r2.navigator.extensions.time
import org.readium.r2.shared.InternalReadiumApi
import org.readium.r2.shared.publication.Locator
import org.readium.r2.shared.util.Url
import kotlin.time.Duration

private const val TAG = "FlutterMediaOverlay"

/**
 * Simple media overlay mapping.
 */
@Parcelize
data class FlutterMediaOverlay(
    val items: List<FlutterMediaOverlayItem>,
) : Parcelable {
    /**
     * The audio file name (without fragment).
     */
    private val audioFile
        get() = items.firstOrNull()?.audioFile ?: ""

    /**
     * The text file name (without fragment).
     */
    @IgnoredOnParcel
    private val textFile
        get() = items.firstOrNull()?.textFile ?: ""

    /**
     * The audio file Url.
     */
    private val audioUrl
        get() = Url.invoke(audioFile)

    /**
     * The text file Url.
     */
    private val textUrl
        get() = Url.invoke(textFile)

    /**
     * The total duration of the audio, based on the end time of the last item.
     */
    val duration
        get() = items.lastOrNull()?.audioEnd ?: 0.0

    /**
     * Find the media overlay item for the given file and time.
     * Returns null if no item is found.
     */
    fun findItemInRange(
        fileHref: Url,
        time: Double,
    ): FlutterMediaOverlayItem? = findItemInRange(fileHref.toString(), time)

    /**
     * Find the media overlay item for the given file and time.
     * Returns null if no item is found.
     */
    fun findItemInRange(
        fileHref: Url,
        time: Duration,
    ): FlutterMediaOverlayItem? = findItemInRange(fileHref, time.inWholeSeconds.toDouble())

    /**
     * Find the media overlay item for the given file and time.
     * Returns null if no item is found.
     */
    fun findItemInRange(
        fileHref: String,
        duration: Duration,
    ): FlutterMediaOverlayItem? = findItemInRange(fileHref, duration.inWholeSeconds.toDouble())

    /**
     * Find the media overlay item for the given file and time.
     * Returns null if no item is found.
     */
    fun findItemInRange(
        fileHref: String,
        time: Double,
    ): FlutterMediaOverlayItem? {
        val href = Url.invoke(fileHref) ?: return null
        if (!href.isEquivalent(textUrl) && !href.isEquivalent(audioUrl)) {
            return null
        }

        return items.firstOrNull { item -> item.isInRange(href, time) }
    }

    /**
     * Find the media overlay item from the text reference.
     */
    fun findItemFromTextId(
        href: Url,
        textId: String,
    ): FlutterMediaOverlayItem? {
        if (!href.isEquivalent(textUrl) && !href.isEquivalent(audioUrl)) {
            return null
        }

        return items.firstOrNull { item -> item.textId == textId }
    }

    /**
     * Find the media overlay item from the given locator.
     * A locator can either be an audio+time based locator or a text+id based locator.
     * This allows us to map back and forth between audio and text.
     *
     * [allowResourceFallback] gates the *imprecise* resource-first fallbacks (an
     * unmatched text id, or no id at all on an HTML resource): when true, an uncued
     * anchor maps to the first cue of the resource; when false, returns null so the
     * caller can leave playback untouched. Pass false when audio is already playing
     * in this same resource (cross-resource check is done at the navigator level).
     * See issue #139.
     */
    @OptIn(InternalReadiumApi::class)
    fun findItemFromLocator(locator: Locator, allowResourceFallback: Boolean = true): FlutterMediaOverlayItem? {
        val href = locator.href
        if (!href.isEquivalent(Url.invoke(textFile)) && !href.isEquivalent(Url.invoke(audioFile))) {
            return null
        }

        locator.locations.time?.let { timeOffset ->
            return findItemInRange(href, timeOffset)
        }

        // DiViNa comic page: the page locator's only anchor is a generic "img" css selector
        // (which getTextId() would otherwise pick up), not a media-overlay text id. Match the
        // page's first narrated item by href — the textref is the image itself.
        if (locator.mediaType.isBitmap) {
            return items.firstOrNull { item -> item.textFile == href.path }
        }

        // Reflowable text: try exact DOM element id match first; fall through on no match.
        locator.getTextId()?.let { textId ->
            findItemFromTextId(href, textId)?.let { return it }
            PluginLog.d(TAG, "::findItemFromLocator - textId '$textId' matched no cue for href=${href.path}")
        }

        locator.progression?.let { progression ->
            val item = items.firstOrNull { item -> item.isInProgression(href, progression) }

            // FIXME: This item?skipToAudioLocator will have an incorrect time value, since it is the original audioStart and not calculated from progression.
            return item
        }

        if (allowResourceFallback && locator.mediaType.isHtml) {
            // No cue matched — fall back to first item of the resource (covers both
            // no-fragment HTML and an id that has no narration entry e.g. a heading).
            PluginLog.d(
                TAG,
                "::findItemFromLocator - resource-fallback: first item for href=${href.path}",
            )
            return items.firstOrNull { item ->
                item.textFile == href.path
            }
        }

        PluginLog.d(
            TAG,
            "::findItemFromLocator - no match for locator=$locator",
        )

        return null
    }

    companion object {
        fun fromJson(
            json: JSONObject,
            position: Int,
            tocHref: Url?,
            title: String,
            readiumOrderItemDuration: Double,
        ): FlutterMediaOverlay? {
            val topNarration = json.opt("narration") as? JSONArray ?: return null
            val items = mutableListOf<FlutterMediaOverlayItem>()
            for (i in 0 until topNarration.length()) {
                val itemJson = topNarration.getJSONObject(i)
                FlutterMediaOverlayItem
                    .fromJson(
                        itemJson,
                        position,
                        tocHref,
                        title,
                        readiumOrderItemDuration,
                    )?.let { items.add(it) }

                fromJson(
                    itemJson,
                    position,
                    tocHref,
                    title,
                    readiumOrderItemDuration,
                )?.let { items.addAll(it.items) }
            }

            return FlutterMediaOverlay(items)
        }
    }
}
